src/main/java/org/fenixedu/academic/ui/rest/PhotographController.java

```diff
@@ -67,34 +64,39 @@ public Response get(@PathParam("username") String username, @QueryParam(value =
             final Photograph personalPhoto =
                     user.getPerson().isPhotoAvailableToCurrentUser() ? user.getPerson().getPersonalPhoto() : null;
 
-            EntityTag entityTag =
-                    new EntityTag(personalPhoto == null ? "mm-av" : personalPhoto.getExternalId() + "-" + size, true);
-            if (!StringUtils.isBlank(ifNoneMatch)
-                    && entityTag.getValue().equals(ifNoneMatch.substring(3, ifNoneMatch.length() - 1))) {
-                return Response.status(Response.Status.NOT_MODIFIED).build();
+            if (personalPhoto != null) {
```

Comment:
@francisco-esteves I think we can refactor this code a bit to take into account a few things:

1. Reduce the cognitive load in the main method. We can have a single return point for buildResponse with a few tweaks.
2. Avoid extra computation such as the resize of the photo if we will not serve it due to entity tag showing that it's still everything up to date.

So my suggestion to achieve this is is using an Optional for the photograph that is passed to the buildRespnse method
something like:

```
   User user = User.findByUsername(username);
   // Notice here that the "trick" is actually to apply a filter to a single object avoiding the
   // ternary operator
   Optional<Photograph> photograph = Optional.ofNullable(user).map(u -> u.getPerson())
                .filter(p -> p.isPhotoAvailableToCurrentUser()).map(p -> p.getPersonalPhoto());

    return buildResponse(ifNoneMatch, size, photograph);
 ```

Then we can write the buildResponse something like:

```
 private Response buildResponse(String ifNoneMatch,  int size, Optional<Photograph> photograph) {
        Response.ResponseBuilder responseBuilder;

        // only compute the entity tag because we might just return if 
       // there's no changes
        EntityTag entityTag = photograph.map(p -> new EntityTag(p.getExternalId() + "-" + size, true))
                .orElseGet(() -> new EntityTag("mm-av", true));

        if (!StringUtils.isBlank(ifNoneMatch)
                && entityTag.getValue().equals(ifNoneMatch.substring(3, ifNoneMatch.length() - 1))) {
            responseBuilder = Response.status(Response.Status.NOT_MODIFIED);
        } else {
            CacheControl cacheControl = new CacheControl();
            cacheControl.setMaxAge(1209600);

            // ChronoUnit.WEEKS was not supported so I used 14 days
            responseBuilder = Response.status(Response.Status.OK).expires(Date.from(Instant.now().plus(14, ChronoUnit.DAYS)))
                    .cacheControl(cacheControl);

// we compute the rest only here
            String mimeType = photograph.map(p -> p.getOriginal().getPictureFileFormat().getMimeType()).orElse("image/png");

            byte[] entity = photograph.map(p -> p.getCustomAvatar(size, size, PictureMode.ZOOM)).orElseGet(() -> {
                try (InputStream mysteryMan = PhotographController.class.getClassLoader()
                        .getResourceAsStream("META-INF/resources/img/mysteryman.png")) {
                    return Avatar.process(mysteryMan, mimeType, size);
                } catch (IOException e) {
                    return new byte[] {};
                }
            });

            responseBuilder.type(mimeType);
            responseBuilder.entity(entity);
            responseBuilder.tag(entityTag);
        }

        return responseBuilder.build();
    }
```